File: 2023/Leetcode/problem202.js
/**
 * @param {number} n
 * @return {boolean}
 */

/*
O(1) - Constant (Fastest)
O(log n) - Logarithmetic
O(n) - Linear 
O(n log n) - Fastest Comparison Sort
O(n^2) - Quadratic
*/

/*
Two pointers (Floyd's Algorithm)

n: Cycle Size
Time Complexity: O(n)
Space Complexity: O(1)
*/

/*
Hash Table (Lookup Table) (Hash Map)

n: Cycle Size
Time Complexity: O(n)
Space Complexity: O(n)
*/

var isHappy = function(n) {
    let mySet = new Set();
    let strNum;
    let mySum;

    while (n != 1) {
        strNum = n.toString();

        mySum = 0;
        for (let i = 0; i < strNum.length; i++) {
            mySum += parseInt(strNum[i]) * parseInt(strNum[i]);
        }

        if (mySet.has(mySum)) {
            return false;
        } else {
            mySet.add(mySum);
            n = mySum;
        }
    }

    return true;
};

File: 2023/Leetcode/problem202.py
class Solution:
    def nextNum(self, n):
        result = 0
        for char in str(n):
            result += int(char) ** 2
        return result

    def isHappy(self, n: int) -> bool:
        mySet = set()

        while n != 1 and n not in mySet:
            mySet.add(n)
            n = self.nextNum(n)
        
        return n == 1